### ISSUE TITLE
Implement file reference counting for copies

### ISSUE THREAD

**Issue by dleehr** [Duke-Translational-Bioinformatics/duke-data-service#646]
During today's GCB/Duke Data Service meeting, we discussed performance and complexity of copying projects, and files in Duke Data Service. At @mgardnerpsu's suggestion, I'm documenting our discussion (and proposal) in a GitHub issue.

[DukeDSClient](https://github.com/Duke-GCB/DukeDSClient) copies projects by downloading them and then uploading them. This is simple, but not optimal. The effective throughput for the copy operation is  between 50-100MB/sec. It works today without any architectural changes to the data service, because the process simply creates a new project and uploads files to it. It's inelegant and failures would happen if the client's network connection is interrupted.

We explored object copy using [Swift's copyObject API](http://developer.openstack.org/api-ref-objectstorage-v1.html#copyObject) to copy the objects. We saw an increase to 220MB/sec, provided we ran 8 operations in parallel (copying the chunks of a SLO). This is better, but is not a drop-in change. It would require implementation in the data-service, perhaps in a task queue, and possibly data model changes.

A third option for copying data (and what I'm proposing) is to forego the literal object copy, and instead implement reference counting semantics. Rather than creating another identical data object in the object store and another File object, a second File object could be created, referencing the same data object in the object store. Copies would be nearly instantaneous for files of any size. As I understand it, the data objects managed are never mutated, and access is always brokered by the data service. This would accomplish copy-on-write, which seems to fit the rules that are already in-place

We briefly discussed some of the implications (billing/responsibility, storage, deletion, de-duplication) too. Costs could be charged to users for every copy of the data, even if there is only one physical copy. Deletion from the object store (which is not currently implemented) would be triggered when the reference count for an object drops to zero. Finally, deduplication could match existing objects based on hashes and treat them as a reference increment. This would have the benefit of maintaining the amount of physical storage used when a user uploads a file that already exists in the data service.

**Comment by Donmurry** [comment]
@dleehr is this still an issue since our latest release with the project transfer endpoints?

**Comment by dleehr** [comment]
Yes it is. To deliver a copy of a project, we download the project to the local filesystem, upload it with a different name, and then transfer the copied project. So #577 is no longer relevant, but this, as a feature request, still stands.
